README.md
# Comp-Arc-HW3

<br> Вариант 1.
<br> Работа выполнена Татауровым Матвеем, БПИ 217 на оценку 6.
<br> Так как в данной задаче примененялось разложение в ряд Маклорена, то краевые значения х установлены как -1 и 1.
<br> Также из-за особенностей данного разложения значения в точках, близким к краевым, ведут себя немного странно, но разобраться почему я так и не смог

## На 4 балла:
 ### 1. Приведено решение задачи на С:
   > * [main.c](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/c-files/main.c)
   > * [countRoot.c](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/c-files/countRoot.c)
### 2. Программа проассемблирована и откомпилирована без оптимизирующих опций, в коде программы прокомментированы переменные:
   > * [main.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/main.s)
   > * [countRoot.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/countRoot.s)
    
   ### <br> Вот файлик со всеми переменными для удобства:
   > * [Переменные](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/other/Variables.md)
   
### 3. Программа промодифицирована флагами компиляции, затем изменена вручную:
   > * [main_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/main_mod.s)
   > * [countRoot_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/countRoot_mod.s)
   ### Файл изменений:
   > * [Список изменений](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/other/Changes.md)
 
 ### 4. Проведено тестирование, результаты указаны в [документе](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/other/tests.md).
## На 5 баллов:
 ### 1. В приведённом выше решении задачи на C используются локальные переменные, а так же 2 функции c параметрами.
   > * [main.c](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/c-files/main.c)
   > * [countRoot.c](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/c-files/countRoot.c)
 ### 2. В ассемблерный код были добавлены комментарии, поясняющие передачу параметров в метод, а так же связь параметров метода с регистрами.
   > * [main_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/main_mod.s)
   > * [countRoot_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/countRoot_mod.s)
   
## На 6 баллов:
### 1. 1.Рефакторинг программы за счет максимального использования регистров процессора:
<br> Получилось произвести лишь одну замену - r12d вместо итераторов i в обоих for

   > * [main_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/main_mod.s)
   > * [countRoot_mod.s](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/asm-files/countRoot_mod.s)
### 2. Проведено тестирование, результаты указаны в [документе](https://github.com/KcasTischaWattt/Comp-Arc-HW3/blob/main/other/tests.md).

### 3. Так как замеров времени данной оценкой не предусмотрено, я могу лишь сравнить размеры исполняемых файлов и количество строк.
Размер исполняемого файла:
>     Программа на С: 16,5 KiB
>     Ассемблерная программа без модификаций: 16,5 KiB
>     Финальная ассемблерная программа с модификациями: 16,3 KiB

Количество строк в файлах суммарно:
>     Программа на С: 14+19 = 33
>     Ассемблерная программа без модификаций: 98+125 = 223
>     Финальная ассемблерная программа с модификациями: 61+90 = 151


other/Changes.md
# Модификация

Код был модифицирован про помощи команды
>     gcc -masm=intel 
>     -fno-asynchronous-unwind-tables
>     -fno-jump-tables
>     -fno-stack-protector 
>     -fno-exceptions 
>     -fverbose-asm 
>     ./file.c -S -o ./file_mod.s
После чего все файлы были отредактированы вручную:

1. Из всех файлов удалены cdqe;
2. Из всех файлов удалены все строки после 
   ```.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0```
   <br> за ненадобностью;
3. Удалены лишние переприсваивания - они перечислены в конце файла.

## Удалённые переприсваивания
### main_mod.s
<br> Было:
``` assembly
  movq	rax, xmm0
	mov	QWORD PTR -8[rbp], rax
	mov	rax, QWORD PTR -8[rbp]
	movq	xmm0, rax
```
<br>Стало:
``` assembly
  movq   QWORD PTR -8[rbp], xmm0
```


asm-files/main.s
	.file	"Main.c"
	.intel_syntax noprefix
	.text
	.section	.rodata
	.align 8
.LC1:
	.string	"Input your number(-1 < x < 1):"
.LC2:
	.string	"%lf"
.LC5:
	.string	"Result: %.30f"
	.text
	.globl	main
	.type	main, @function
main:
.LFB0:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	sub	rsp, 32
	mov	rax, QWORD PTR fs:40
	mov	QWORD PTR -8[rbp], rax
	xor	eax, eax
	movsd	xmm0, QWORD PTR .LC0[rip]
	movsd	QWORD PTR -24[rbp], xmm0     # double x
.L2:
	lea	rdi, .LC1[rip]
	call	puts@PLT
	lea	rax, -24[rbp]
	mov	rsi, rax
	lea	rdi, .LC2[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	movsd	xmm0, QWORD PTR -24[rbp]
	movsd	xmm1, QWORD PTR .LC3[rip]
	comisd	xmm0, xmm1
	jnb	.L2
	movsd	xmm1, QWORD PTR -24[rbp]
	movsd	xmm0, QWORD PTR .LC4[rip]
	comisd	xmm0, xmm1
	jnb	.L2
	mov	rax, QWORD PTR -24[rbp]
	movq	xmm0, rax
	call	countRoot@PLT
	movq	rax, xmm0
	mov	QWORD PTR -16[rbp], rax     # double res
	mov	rax, QWORD PTR -16[rbp]
	movq	xmm0, rax
	lea	rdi, .LC5[rip]
	mov	eax, 1
	call	printf@PLT
	mov	eax, 0
	mov	rdx, QWORD PTR -8[rbp]
	xor	rdx, QWORD PTR fs:40
	je	.L4
	call	__stack_chk_fail@PLT
.L4:
	leave
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	main, .-main
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1073741824
	.align 8
.LC3:
	.long	0
	.long	1072693248
	.align 8
.LC4:
	.long	0
	.long	-1074790400
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:

asm-files/main_mod.s
	.file	"Main.c"
	.intel_syntax noprefix
	.text
	.section	.rodata
	.align 8
.LC1:
	.string	"Input your number(-1 < x < 1):"
.LC2:
	.string	"%lf"
.LC5:
	.string	"Result: %.30f"
	.text
	.globl	main
	.type	main, @function
main:
	endbr64
	push	rbp
	mov	rbp, rsp
	sub	rsp, 16
	movsd	xmm0, QWORD PTR .LC0[rip]
	movsd	QWORD PTR -16[rbp], xmm0
.L2:
	lea	rdi, .LC1[rip]
	call	puts@PLT
	lea	rax, -16[rbp]
	mov	rsi, rax
	lea	rdi, .LC2[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	movsd	xmm0, QWORD PTR -16[rbp]
	movsd	xmm1, QWORD PTR .LC3[rip]
	comisd	xmm0, xmm1
	jnb	.L2
	movsd	xmm1, QWORD PTR -16[rbp]
	movsd	xmm0, QWORD PTR .LC4[rip]
	comisd	xmm0, xmm1
	jnb	.L2
	mov	rax, QWORD PTR -16[rbp] # Передача x в функцию
	movq	xmm0, rax
	call	countRoot@PLT
	movq   QWORD PTR -8[rbp], xmm0
	lea	rdi, .LC5[rip]
	mov	eax, 1
	call	printf@PLT
	mov	eax, 0
	leave
	ret
	.size	main, .-main
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1073741824
	.align 8
.LC3:
	.long	0
	.long	1072693248
	.align 8
.LC4:
	.long	0
	.long	-1074790400

asm-files/countRoot_mod.s
	.file	"countRoot.c"
	.intel_syntax noprefix
	.text
	.globl	pow
	.type	pow, @function
pow:
	endbr64
	push	rbp
	mov	rbp, rsp
	movsd	QWORD PTR -24[rbp], xmm0 # Первый параметр - double x
	movsd	QWORD PTR -32[rbp], xmm1 # Второй параметр - double y
	movsd	xmm0, QWORD PTR -24[rbp]
	movsd	QWORD PTR -8[rbp], xmm0
	mov	r12d, 1
	jmp	.L2
.L3:
	movsd	xmm0, QWORD PTR -8[rbp]
	mulsd	xmm0, QWORD PTR -24[rbp]
	movsd	QWORD PTR -8[rbp], xmm0
	add	r12d, 1
.L2:
	cvtsi2sd	xmm1, r12d
	movsd	xmm0, QWORD PTR -32[rbp]
	comisd	xmm0, xmm1
	ja	.L3
	movsd	xmm0, QWORD PTR -8[rbp] # return x
	pop	rbp
	ret
	.size	pow, .-pow
	.globl	countRoot
	.type	countRoot, @function
countRoot:
	endbr64
	push	rbp
	mov	rbp, rsp
	sub	rsp, 40
	movsd	QWORD PTR -40[rbp], xmm0 # Первый параметр - double x
	movsd	xmm1, QWORD PTR -40[rbp]
	movsd	xmm0, QWORD PTR .LC0[rip]
	mulsd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC1[rip]
	addsd	xmm0, xmm1
	movsd	QWORD PTR -8[rbp], xmm0
	movsd	xmm0, QWORD PTR .LC0[rip]
	movsd	QWORD PTR -16[rbp], xmm0
	mov	QWORD PTR -24[rbp], 1
	mov	r12d, 2
	jmp	.L6
.L7:
	cvtsi2sd	xmm1, r12d
	movsd	xmm0, QWORD PTR .LC0[rip]
	subsd	xmm0, xmm1
	movapd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC1[rip]
	addsd	xmm0, xmm1
	movsd	xmm1, QWORD PTR -16[rbp]
	mulsd	xmm0, xmm1
	movsd	QWORD PTR -16[rbp], xmm0
	mov	eax, r12d
	mov	rdx, QWORD PTR -24[rbp]
	imul	rax, rdx
	mov	QWORD PTR -24[rbp], rax 
	cvtsi2sd	xmm0, r12d
	mov	rax, QWORD PTR -40[rbp] # Передача double x в pow
	movapd	xmm1, xmm0	# Передача i в pow
	movq	xmm0, rax
	call	pow
	mulsd	xmm0, QWORD PTR -16[rbp]
	cvtsi2sd	xmm1, QWORD PTR -24[rbp]
	divsd	xmm0, xmm1
	movsd	xmm1, QWORD PTR -8[rbp]
	addsd	xmm0, xmm1
	movsd	QWORD PTR -8[rbp], xmm0
	add	r12d, 1
.L6:
	cmp	r12d, 14
	jle	.L7
	movsd	xmm0, QWORD PTR -8[rbp] # return res
	leave
	ret
	.size	countRoot, .-countRoot
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1071644672
	.align 8
.LC1:
	.long	0
	.long	1072693248


asm-files/countRoot.s
	.file	"countRoot.c"
	.intel_syntax noprefix
	.text
	.globl	pow
	.type	pow, @function
pow:
.LFB0:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	movsd	QWORD PTR -24[rbp], xmm0 # double x
	movsd	QWORD PTR -32[rbp], xmm1 # double y
	movsd	xmm0, QWORD PTR -24[rbp]
	movsd	QWORD PTR -8[rbp], xmm0 # double temp
	mov	DWORD PTR -12[rbp], 1 # int i в for(в функции pow)
	jmp	.L2
.L3:
	movsd	xmm0, QWORD PTR -8[rbp]
	mulsd	xmm0, QWORD PTR -24[rbp]
	movsd	QWORD PTR -8[rbp], xmm0
	add	DWORD PTR -12[rbp], 1
.L2:
	cvtsi2sd	xmm1, DWORD PTR -12[rbp]
	movsd	xmm0, QWORD PTR -32[rbp]
	comisd	xmm0, xmm1
	ja	.L3
	movsd	xmm0, QWORD PTR -8[rbp]
	pop	rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	pow, .-pow
	.globl	countRoot
	.type	countRoot, @function
countRoot:
.LFB1:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	sub	rsp, 40
	movsd	QWORD PTR -40[rbp], xmm0 # double x
	movsd	xmm1, QWORD PTR -40[rbp]
	movsd	xmm0, QWORD PTR .LC0[rip]
	mulsd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC1[rip]
	addsd	xmm0, xmm1
	movsd	QWORD PTR -24[rbp], xmm0 # double res
	movsd	xmm0, QWORD PTR .LC0[rip]
	movsd	QWORD PTR -16[rbp], xmm0 # double alpha
	mov	QWORD PTR -8[rbp], 1 # long long factorial
	mov	DWORD PTR -28[rbp], 2 # int i в for(в функции countRoot)
	jmp	.L6
.L7:
	cvtsi2sd	xmm1, DWORD PTR -28[rbp]
	movsd	xmm0, QWORD PTR .LC0[rip]
	subsd	xmm0, xmm1
	movapd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC1[rip]
	addsd	xmm0, xmm1
	movsd	xmm1, QWORD PTR -16[rbp]
	mulsd	xmm0, xmm1
	movsd	QWORD PTR -16[rbp], xmm0
	mov	eax, DWORD PTR -28[rbp]
	cdqe
	mov	rdx, QWORD PTR -8[rbp]
	imul	rax, rdx
	mov	QWORD PTR -8[rbp], rax
	cvtsi2sd	xmm0, DWORD PTR -28[rbp]
	mov	rax, QWORD PTR -40[rbp]
	movapd	xmm1, xmm0
	movq	xmm0, rax
	call	pow
	mulsd	xmm0, QWORD PTR -16[rbp]
	cvtsi2sd	xmm1, QWORD PTR -8[rbp]
	divsd	xmm0, xmm1
	movsd	xmm1, QWORD PTR -24[rbp]
	addsd	xmm0, xmm1
	movsd	QWORD PTR -24[rbp], xmm0
	add	DWORD PTR -28[rbp], 1
.L6:
	cmp	DWORD PTR -28[rbp], 14
	jle	.L7
	movsd	xmm0, QWORD PTR -24[rbp]
	leave
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE1:
	.size	countRoot, .-countRoot
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1071644672
	.align 8
.LC1:
	.long	0
	.long	1072693248
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4: